sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/source/CosmosSourceTask.java

```diff
@@ -0,0 +1,333 @@
+        CosmosAsyncContainer container =
+            this.cosmosClient
+                .getDatabase(feedRangeTaskUnit.getDatabaseName())
+                .getContainer(feedRangeTaskUnit.getContainerName());
+        return asyncDocumentClient
+            .getCollectionCache()
+            .resolveByNameAsync(null, BridgeInternal.extractContainerSelfLink(container), null)
+            .flatMap(collection -> {
+                return asyncDocumentClient.getPartitionKeyRangeCache().tryGetOverlappingRangesAsync(
+                    null,
+                    collection.getResourceId(),
+                    feedRangeTaskUnit.getFeedRange(),
+                    true,
+                    null);
+            })
+            .flatMap(pkRangesValueHolder -> {
+                if (pkRangesValueHolder == null || pkRangesValueHolder.v == null) {
+                    return Mono.error(new IllegalStateException("There are no overlapping ranges for the range"));
+                }
+
+                List<PartitionKeyRange> partitionKeyRanges = pkRangesValueHolder.v;
+                if (partitionKeyRanges.size() == 1) {
+                    // merge happens
+                    logger.info(
+                        "FeedRange {} is merged into {}, but we will continue polling data from feedRange {}",
+                        feedRangeTaskUnit.getFeedRange(),
+                        partitionKeyRanges.get(0).toRange(),
+                        feedRangeTaskUnit.getFeedRange());
+
+                    // Continue using polling data from the current task unit feedRange
+                    return Mono.just(false);
+                } else {
+                    logger.info(
+                        "FeedRange {} is split into {}. Will create new task units. ",
+                        feedRangeTaskUnit.getFeedRange(),
+                        partitionKeyRanges.stream().map(PartitionKeyRange::toRange).collect(Collectors.toList())
+                    );
+
+                    for (PartitionKeyRange pkRange : partitionKeyRanges) {
+                        FeedRangeTaskUnit childTaskUnit =
+                            new FeedRangeTaskUnit(
+                                feedRangeTaskUnit.getDatabaseName(),
+                                feedRangeTaskUnit.getContainerName(),
+                                feedRangeTaskUnit.getContainerRid(),
+                                pkRange.toRange(),
+                                feedRangeTaskUnit.getContinuationState(),
+                                feedRangeTaskUnit.getTopic());
+                        this.taskUnitsQueue.add(childTaskUnit);
+                    }
+
+                    // remove the current task unit from the queue
+                    return Mono.just(true);
+                }
+            });
+    }
+
+    private String getItemLsn(JsonNode item) {
+        return item.get(LSN_ATTRIBUTE_NAME).asText();
```

Comment:
I think there are weird cases where LSN comes as plain number vs. "<number>" - check Spark code - we have some maginc in there to be able to parse both versions correctly

Reply:
whether it is number node or text node, using asText should be safe

Reply:
Discussed offline, the isn may in different format depends on where we extract the lsn. For example, it will be "\"LSN\"" in continuationToken, while in "LSN" format form headers. 

The LSN being extracted here is from the item objectNode, which should be in "LSN" format.